core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1470,21 +1503,45 @@ class KafkaApis(val requestChannel: RequestChannel,
     val transactionalId = addPartitionsToTxnRequest.transactionalId
     val partitionsToAdd = addPartitionsToTxnRequest.partitions
 
-    // Send response callback
-    def sendResponseCallback(error: Errors): Unit = {
-      def createResponse(throttleTimeMs: Int): AbstractResponse = {
-        val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(throttleTimeMs, error)
-        trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-        responseBody
+    if(!authorize(request.session, Write, new Resource(ProducerTransactionalId, transactionalId)))
+      sendResponseMaybeThrottle(request, (throttleTimeMs: Int) => new InitPidResponse(throttleTimeMs, Errors.TRANSACTIONAL_ID_AUTHORIZATION_FAILED))
+    else {
+      val internalTopics = partitionsToAdd.asScala.filter {tp => kafka.common.Topic.InternalTopics.contains(tp.topic())}
+
+      val (existingAndAuthorizedForDescribeTopics, nonExistingOrUnauthorizedForDescribeTopics) =
```

Comment:
I'm not really sure how we want to do this. Most of the existing errors for this are at the request level, i.e., `NotCoordinator`, `InvalidPidMapping`, etc. 
But then we have the topic level errors, too. Do we keep the request level error and add an additional Map for any topic level errors?